TFLu: Port TFL detection postprocess operator

mansnils opened the pull request.

mansnils committed:
TFLu: Port TFL detection postprocess operator

Change-Id: I9edffb31a3b1485ebc43e2298762ad067bd26dda

njeffrie reviewed (changes requested):
Update: upon closer inspection, it looks like std::vector is used a lot here. We avoid dynamic memory allocation on TF Micro, so converting to c style arrays or the TF Micro memory API will be necessary.

mansnils committed:
TFlu: replace std::vector in detection_postprocess

mansnils commented:
@njeffrie Right, it is done.

njeffrie reviewed (dismissed):
Thanks.  I'll have to run this internally again and hopefully we can land it.

mansnils committed:
TFlu: Fix detection_postprocess output dimension handling

Tensors with undefined dimension will not really have it undefined
as the actual flatbuffer tensor. Instead the dimension size will be zero.

mansnils commented:
@njeffrie I realized that the output dimension handling needed an update.

rthadur commented on tensorflow/lite/micro/kernels/detection_postprocess.cc:
```diff
@@ -0,0 +1,857 @@
+}
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  // TODO(chowdhery): Generalize for any batch size
```
please remove todo

rthadur commented on tensorflow/lite/micro/kernels/detection_postprocess.cc:
```diff
@@ -0,0 +1,857 @@
+  // highest scoring non-overlapping boxes.
+  TF_LITE_ENSURE_STATUS(NonMaxSuppressionMultiClass(context, node, op_data));
+
+  // TODO(chowdhery): Generalize for any batch size
```
same here

rthadur commented:
@mansnils can you please check below errors as well 

`tensorflow/lite/micro/kernels/detection_postprocess.cc:18:10: error: module ///tensorflow/lite/micro/kernels:micro_ops does not depend on a module exporting '/flatbuffers/include/flatbuffers/flexbuffers.h'

build_cleaner ///tensorflow/lite/micro/kernels:micro_ops
#include "/flatbuffers/include/flatbuffers/flexbuffers.h"
         ^/tensorflow/lite/micro/kernels/detection_postprocess.cc:859:66: error: missing field 'profiling_string' initializer [-Werror,-Wmissing-field-initializers]
      detection_postprocess::Prepare, detection_postprocess::Eval};`

mansnils committed:
TFLu: detection_postprocess: fix review comments and build issues

mansnils commented:
@rthadur build errors and review comments resolved

rthadur commented on tensorflow/lite/micro/kernels/detection_postprocess.cc:
```diff
@@ -0,0 +1,857 @@
+#include <numeric>
+
+#define FLATBUFFERS_LOCALE_INDEPENDENT 0
+#include "flatbuffers/flexbuffers.h"  // from @flatbuffers
```
`third_party/tensorflow/lite/micro/kernels/detection_postprocess.cc:18:10: error: module //third_party/tensorflow/lite/micro/kernels:micro_ops does not depend on a module exporting 'third_party/flatbuffers/include/flatbuffers/flexbuffers.h'
see http://go/cpp-features#layering_check; to fix run:
build_cleaner //third_party/tensorflow/lite/micro/kernels:micro_ops
#include "third_party/flatbuffers/include/flatbuffers/flexbuffers.h"`

rthadur reviewed (changes requested):
@mansnils can you please check below review comments ?

gbaned commented:
@mansnils Can you please check @rthadur's comments and keep us posted ? Thanks!

gbaned commented:
@mansnils, Any update on this PR? Please. Thanks!

mansnils committed:
TFLu: add missing bazel dependency for detection_pp

mansnils commented:
@rthadur Comment and conflicts resolved.

rthadur commented:
```
tensorflow/lite/micro/kernels/detection_postprocess.cc:288:20: error: unknown type name 'uint8'; did you mean 'uint'?
  float operator()(uint8 x) 

third_party/tensorflow/lite/micro/kernels/detection_postprocess.cc:301:9: error: unknown type name 'uint8'; did you mean 'uint'

third_party/tensorflow/lite/micro/kernels/detection_postprocess.cc:302:21: error: unknown type name 'uint8'; did you mean 'uint'

third_party/tensorflow/lite/micro/kernels/detection_postprocess.cc:746:9: error: unknown type name 'uint8'; did you mean 'uint'

third_party/tensorflow/lite/micro/kernels/detection_postprocess.cc:746:45: error: unknown type name 'uint8'; did you mean 'uint'
```

@mansnils can you please check this errors ?

mansnils committed:
TFLu: Fix for internal build error in detection_pp

mansnils commented:
@rthadur errors resolved.

rthadur commented:
@mansnils can you please check below errors 

```
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:19:10: error: module ///tensorflow/lite/micro/kernels:detection_postprocess_test_binary does not depend on a module exporting '/tensorflow/lite/micro/all_ops_resolver.h'
build_cleaner ///tensorflow/lite/micro/kernels:detection_postprocess_test_binary
#include "third_party/tensorflow/lite/micro/all_ops_resolver.h"
         ^
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:125:17: error: unused variable 'input_scale2' 
    const float input_scale2 = ScaleFromMinMax<uint8_t>(input_min2, input_max2);
                ^
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:263:13: error: unused variable 'kInputElements2' 
  const int kInputElements2 = tflite::testing::kInputShape2[1] *
            ^
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:266:13: error: unused variable 'kInputElements3' 
  const int kInputElements3 =
            ^
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:319:13: error: unused variable 'kInputElements2' 
  const int kInputElements2 = tflite::testing::kInputShape2[1] *
            ^
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:322:13: error: unused variable 'kInputElements3'
  const int kInputElements3 =
            ^
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:470:13: error: unused variable 'kInputElements3' 
  const int kInputElements3 =
            ^
/tensorflow/lite/micro/kernels/detection_postprocess_test.cc:467:13: error: unused variable 'kInputElements2'
  const int kInputElements2 = tflite::testing::kInputShape2[1] *
```

mansnils committed:
TFLu: Fix warnings in detection_pp test

mansnils commented:
@rthadur errors resolved

njeffrie commented:
Apologies for the repeated back-and-forth Måns - we have an internal version of this CL now and there are a few more things we will need before we can land this:

- We recently switched to TfLiteEvalTensor instead of TfLiteTensor during invoke stage.  Init and prepare can still use the TfLiteTensor, but they are de-allocated at invoke for significant memory savings. See https://github.com/tensorflow/tensorflow/commit/88461053262f02bbc15887daa172c02db7419780 for how we ported the cmsis-nn kernels over to this new kernel API.

- It seems like there are some unused variables which are causing compiler errors. Can you try to compile with -Werror=unused-variable?

I left a few inline comments on the change as well.

Thanks and sorry again for the repeated back-and-forth.

njeffrie commented on tensorflow/lite/micro/kernels/detection_postprocess.cc:
```diff
@@ -0,0 +1,855 @@
+  const int num_classes = op_data->num_classes;
+
+  // Scratch tensors
+  context->RequestScratchBufferInArena(context, num_boxes,
```
RequestScratchBufferInArena will request a temporary buffer which will be gone at the end of the given stage. This means requesting scratch buffers in prepare and referencing them from OpData is invalid, since they will have been deallocated before calls to invoke. Please use
buffer = context->AllocatePersistentBuffer(
          context, size_in_bytes). See an example here https://github.com/tensorflow/tensorflow/blob/7ca0373ad3bfeec067c9c604c0973f8455473fdc/tensorflow/lite/micro/kernels/conv.cc#L142

mansnils replied:
I think it is only scratch tensors and temporary buffers (to get rid of std::vector) that are using RequestScratchBufferInArena. And I don't think they need to be preserved at the end of given stage. I have added scratch indexes (*_idx) in OpData as is needed when using RequestScratchBufferInArena. 
I have also added pointers in OpData for these buffers that are set in Eval. The only reason for this is to reduce the number of parameters. Apart from that it is not really needed. It will also look more similar to the TFLite version. However they could be removed. I added some comments to clarify this. Please let me know what you think.

njeffrie replied:
Good call on the scope.  I incorrectly thought they needed to be persistent.

njeffrie commented on tensorflow/lite/micro/kernels/detection_postprocess_test.cc:
```diff
@@ -0,0 +1,482 @@
+
+  // Instance of a zero-length int to pass as tensor dims for a flatbuffer
+  // Tensor with no shape.
+  const TfLiteIntArray kZeroLengthIntArray = {0};
```
Please create int array and use IntArrayFromInts as follows:
const int zero_length_int_array_data[] = {0};
TfLiteIntArray zero_length_int_array = IntArrayFromInts(zero_length_int_array_data);

mansnils replied:
Done

njeffrie commented on tensorflow/lite/micro/kernels/detection_postprocess_test.cc:
```diff
@@ -0,0 +1,482 @@
+  node.builtin_data = nullptr;
+  node.custom_initial_data = nullptr;
+  node.custom_initial_data_size = 0;
+  node.delegate = nullptr;
```
delegate has been removed from TF Micro

mansnils replied:
Done

njeffrie commented on tensorflow/lite/micro/kernels/detection_postprocess_test.cc:
```diff
@@ -0,0 +1,482 @@
+  TfLiteNode node;
+  node.inputs = inputs_array;
+  node.outputs = outputs_array;
+  node.temporaries = nullptr;
```
temporaries has been removed from TF Micro.

mansnils replied:
Done

mansnils committed:
TFLu: Switch to TFLiteEvalTensor in detection_pp

mansnils commented:
@njeffrie Thanks for the comments. I added -Wunused-variable and -Werror but could not find any more unused variables. I also manually added an unused variable to provoke an error to see that the -W options was getting through.

njeffrie commented on tensorflow/lite/micro/kernels/detection_postprocess.cc:
```diff
@@ -0,0 +1,872 @@
+  int buffer_idx;
+  int selected_idx;
+
+  // These are just temporary pointers to scratch buffers, set in each invocation of eval
```
It's not a big deal one way or another, but i prefer using scratch buffer indices to access scratch buffers wherever necessary instead of maintaining references in OpData. That way it's clearer in the implementation, and we can avoid all of these extra members. A side benefit is a bit less boilerplate code:
T* scratch_buffer = reinterpret_cast<T*>(context->GetScratchBuffer(context, op_data->scratch_buffer_idx));

Accessing scratch buffers based on index is a lightweight operation, and it's OK to access the same shared scratch buffer from multiple places, since we are just indexing an array of pointers.

My comments before may have incorrectly pushed you towards this implementation, so apologies for any confusion I caused.

mansnils replied:
Ok, removing the temporay scratch buffer pointers.

mansnils committed:
TFLu: Remove temporary scratch pointers in OpData in detection_pp

mansnils commented:
@njeffrie Updated after comments.

mansnils commented:
@rthadur gentle ping for merge

rthadur commented:
@mansnils thanks for your patience , still in progress merging internally , will keep you updated ..!

mansnils commented:
@rthadur still merging internally?

rthadur commented:
yes, @njeffrie has been working on merging this internally , @njeffrie gentle ping for update ?
